src/test/integration/DesignSurface/DesignSurfaceExt/Framework/Index.cs

```diff
@@ -0,0 +1,166 @@
+            ThrowValueArgumentOutOfRange_NeedNonNegNumException();
+        }
+
+        return new Index(~value);
+    }
+
+    /// <summary>Returns the index value.</summary>
+    public int Value
+    {
+        get
+        {
+            if (_value < 0)
+                return ~_value;
+            else
+                return _value;
+        }
+    }
+
+    /// <summary>Indicates whether the index is from the start or the end.</summary>
+    public bool IsFromEnd => _value < 0;
+
+    /// <summary>Calculate the offset from the start using the giving collection length.</summary>
+    /// <param name="length">The length of the collection that the Index will be used with. length has to be a positive value</param>
+    /// <remarks>
+    ///  <para>
+    ///   For performance reason, we don't validate the input length parameter and the returned offset value against negative values.
+    ///   We don't validate either the returned offset is greater than the input length.
+    ///   It is expected Index will be used with collections which always have non-negative length/count. If the returned offset is negative and
+    ///   then used to index a collection will get out of range exception which will be the same effect as the validation.
+    ///  </para>
+    /// </remarks>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public int GetOffset(int length)
+    {
+        int offset = _value;
+        if (IsFromEnd)
+        {
+            // offset = length - (~value)
+            // offset = length + (~(~value) + 1)
+            // offset = length + value + 1
+
+            offset += length + 1;
+        }
+
+        return offset;
+    }
+
+    /// <summary>Indicates whether the current Index object is equal to another object of the same type.</summary>
+    /// <param name="value">An object to compare with this object</param>
+    public override bool Equals([NotNullWhen(true)] object? value) => value is Index index && _value == index._value;
+
+    /// <summary>Indicates whether the current Index object is equal to another Index object.</summary>
+    /// <param name="other">An object to compare with this object</param>
+    public bool Equals(Index other) => _value == other._value;
+
+    /// <summary>Returns the hash code for this instance.</summary>
+    public override int GetHashCode() => _value;
+
+    /// <summary>Converts integer number to an Index.</summary>
+    public static implicit operator Index(int value) => FromStart(value);
+
+    /// <summary>Converts the value of the current Index object to its equivalent string representation.</summary>
+    public override string ToString()
+    {
+        if (IsFromEnd)
+            return ToStringFromEnd();
+
+        return ((uint)Value).ToString();
+    }
+
+    private static void ThrowValueArgumentOutOfRange_NeedNonNegNumException()
+    {
+#if SYSTEM_PRIVATE_CORELIB
+            throw new ArgumentOutOfRangeException("value", SR.ArgumentOutOfRange_NeedNonNegNum);
```

Comment:
```suggestion
        throw new ArgumentOutOfRangeException("value", SR.ArgumentOutOfRange_NeedNonNegNum);
```
please fix indentation